Lib9c/Model/State/ShopState.cs

```diff
@@ -9,126 +10,236 @@
+        }
+
+        public bool TryUnregister(
+            Address sellerAgentAddress,
+            Guid productId,
+            out ShopItem unregisteredItem)
+        {
+            if (!_products.ContainsKey(productId))
             {
+                unregisteredItem = null;
                 return false;
             }
 
-            var list = AgentProducts[sellerAgentAddress];
+            var targetShopItem = _products[productId];
+            _products.Remove(productId);
 
-            foreach (var shopItem in list)
+            if (_agentProducts.ContainsKey(sellerAgentAddress))
             {
-                if (shopItem.ProductId != productId)
+                var shopItems = _agentProducts[sellerAgentAddress];
+                if (shopItems.Any(item => item.Equals(productId)))
                 {
-                    continue;
+                    shopItems.Remove(productId);
+                    if (shopItems.Count == 0)
+                    {
+                        _agentProducts.Remove(sellerAgentAddress);
+                    }
                 }
+            }
 
-                outPair = new KeyValuePair<Address, ShopItem>(sellerAgentAddress, shopItem);
-                return true;
+            var itemSubType = targetShopItem.ItemUsable.ItemSubType;
+            if (_itemSubTypeProducts.ContainsKey(itemSubType))
+            {
+                var shopItems = _itemSubTypeProducts[itemSubType];
+                shopItems.Remove(productId);
+                if (shopItems.Count == 0)
+                {
+                    _itemSubTypeProducts.Remove(itemSubType);
+                }
             }
 
-            return false;
+            unregisteredItem = targetShopItem;
+            return true;
         }
 
-        public bool TryUnregister(Address sellerAgentAddress,
-            Guid productId, out ShopItem outUnregisteredItem)
+        #endregion
+
+        public bool TryGet(
+            Address sellerAgentAddress,
+            Guid productId,
+            out ShopItem shopItem)
         {
-            if (!TryGet(sellerAgentAddress, productId, out var outPair))
+            if (!_agentProducts.ContainsKey(sellerAgentAddress))
+            {
+                shopItem = null;
+                return false;
+            }
+
+            var shopItems = _agentProducts[sellerAgentAddress];
+            if (shopItems.All(item => item != productId))
             {
-                outUnregisteredItem = null;
+                shopItem = null;
                 return false;
             }
 
-            AgentProducts[outPair.Key].Remove(outPair.Value);
+            if (!_products.ContainsKey(productId))
+            {
+                shopItem = null;
+                return false;
+            }
 
-            outUnregisteredItem = outPair.Value;
+            shopItem = _products[productId];
             return true;
         }
+    }
+
+    [Serializable]
+    public class ShopStateAlreadyContainsException : Exception
```

Comment:
Can you please move this exception to new file? It's not necessary, but easy to look up in IDEs :)

Reply:
@limebell If you create a ShopStateAlreadyContainsException.cs file, which path would be good?
Is it okay to have the same path as the ShopState.cs file?

Reply:
Yes. Since the exception is called at files in `Lib9c/Model/State/` only, so it's good to have same path with `ShotState.cs`

Reply:
@limebell This has been applied!